config/opal_check_ofi.m4

```diff
@@ -100,6 +101,17 @@ AC_DEFUN([OPAL_CHECK_OFI],[
 	AC_SUBST($1_CPPFLAGS)
 	AC_SUBST($1_LDFLAGS)
 	AC_SUBST($1_LIBS)
+    AC_ARG_ENABLE(mtl-ofi-alternative-tag,
+        AC_HELP_STRING([--enable-mtl-ofi-alternative-tag],
+                       [Use alternative ofi tag for providers that do not support FI_REMOTE_CQ_DATA: 12b com_id,16b source,32b mpi_tag (default: 26b com_id, 18b source, 16b mpi_tag)]))
+    if test "$enable_mtl_ofi_alternative_tag" = "yes"; then
+        ALTERNATIVE_TAG=1
+    else
+        ALTERNATIVE_TAG=0
+    fi
+
+    AC_DEFINE_UNQUOTED([MTL_OFI_ALTERNATIVE_DEFAULT_TAG],$ALTERNATIVE_TAG,
+                   [Use ofi alternative no FI_REMOTE_CQ_DATA tag])
```

Comment:
I don't quite understand what this `configure` CLI option is for.  It sounds like if the provider does not support `FI_REMOTE_CQ_DATA`, you use a different bit mapping scheme for tag matching.  Is that right?

If so, why isn't that just detected and used a run time -- why does it require a configure-time argument?

Reply:
There is indeed a runtime detection of FI_REMOTE_CQ_DATA and fallback to a default OFI tag fields distribution (mtl_ofi_component.c 415 to 458). However, the original implementation offered very few bits of the OFI tag to pack the source rank (16) considerably limiting scalability. @bosilca suggested having 2 options for the default/fallback OFI distribution tag that can be selected at configure time, with the "default default" having more bits for the source rank than for the user tag (as openib btl does). See mtl_ofi_types.h 78-81. This option is to choose DEFAULT_2 at build time.

Reply:
Is there a reason to keep the old / unscalable-number-of-bits-for-source default values?

Should the number of source bits be dynamically determined by the job size / size of MPI_COMM_WORLD?

Reply:
@jsquyres  Would it be possible to merge the PR as is?  We have a few other patches we want to contribute that depend on this one. I understand and agree that a dynamic approach for defining the tag bits is the optimal way to go. However, a) most of the providers today support FI_REMOTE_CQ_DATA b) as George shared, there risk of the default TAG not being enough is very reduced, the build time flag is just an alternative for when that odd case happens. We can indeed revisit the the dynamic approach if we see the concern grows.  Thanks,

Reply:
Honestly, this is effectively your+Cray's module, so if you can get @hppritcha's approval, you can go ahead and merge.  But I still think:

1. Having a compile-time decision like this is somewhat un-OMPI-ish.  IMHO, it would be more natural to have a decision like this be a run-time / MCA-var-driven decision.  But I'm not quite sure what decision you're trying to get from the user -- isn't it all driven by what the underlying libfabric provider provides?  I.e., if the fi provider supports 32 bits in FI_REMOTE_CQ_DATA, then you use scheme A.  If it doesn't, you use scheme B.  Does this need to be a user-driven decision at all?
1. Does the OFI MTL convey back the max supported tag value back up to the CM PML?

I agree that dynamically sizing the number of bits based on the size of the job could be considered outside the scope of this PR.  Random question, though: does the OFI MTL support spawn (i.e., adding more peers)?  If not, then you could consider the job size as constant.

Reply:
@jsquyres 
1) Strictly speaking, it can be made a runtime decision. However, given the low chance of occurrence @bolica suggested making it a build time.  
Clarification of the decision being made by the option:  (also clarifies @bwbarrett 's question):  OMPI needs to send 96 bits with each message (32 source rank, 32 communicator id, 32 user tag) but OFI tag only offers 64. When the OFI provider supports FI_REMOTE_CQ_DATA (detected at runtime) there is no limitation since 32 bit is the minimum required by OFI spec. However, when it is not supported some bits need to be trimmed (96 -> 64). Then is when *this* option comes into play and offers two alternatives for "fallback tag" with different options to distribute the 64 bits available.   

2) yes, this is shown in the comments belwo

>Random question, though: does the OFI MTL support spawn (i.e., adding more peers)?

Yes.